File: lib/model/codename.py
from datetime import datetime

from slugify import slugify
from sqlalchemy import Column, DateTime, Integer, String, Text
from sqlalchemy.orm import relationship

from model import Base
from model.image import Image
from model.reference import Reference

class Codename(Base):
    ''' Data model for a codename. '''

    __tablename__ = 'codename'

    id = Column(Integer, primary_key=True)
    name = Column(String(255), unique=True)
    slug = Column(String(255), unique=True)
    summary = Column(Text)
    description = Column(Text)
    added = Column(DateTime)
    updated = Column(DateTime)

    references = relationship(
        'Reference',
        backref='codename',
        cascade='all,delete-orphan'
    )

    images = relationship(
        'Image',
        backref='codename',
        cascade='all,delete-orphan'
    )

    def __init__(self, name):
        ''' Constructor. '''

        self.name = name
        self.slug = slugify(name)
        self.summary = 'Summary placeholder text.'
        self.description = 'Summary placeholder text.'
        self.added = datetime.today()
        self.updated = datetime.today()

File: static/dart/lib/model/codename.dart
import 'package:nsa_codenames/model/image.dart';
import 'package:nsa_codenames/model/reference.dart';

class Codename {
    String name, slug, summary, description;
    DateTime added, updated;
    List<Image> images;
    List<Reference> references;

    Codename.old(this.name, this.description);

    Codename(Map json) {
        this.name = json['name'];
        this.slug = json['slug'];
        this.summary = json['summary'];
        this.description = json['description'];

        this.added = new DateTime.fromMillisecondsSinceEpoch(json['added']);
        this.updated = new DateTime.fromMillisecondsSinceEpoch(json['updated']);

        this.images = new List<Image>();
        for (Map imageJson in json['images']) {
            this.images.add(new Image(imageJson));
        }

        this.references = new List<Reference>();
        for (Map referenceJson in json['references']) {
            this.references.add(new Reference(referenceJson));
        }
    }
}
